lib/compiler/passes/generate-js.js

```diff
@@ -351,6 +352,50 @@ function generateJS(ast, options) {
         parts.push("}");
       }
 
+      /*
+        MATCH_* opcodes typically do something like
+
+          if (<test>(input.substr(peg$currPos, length))) {
+            sN = input.substr(peg$currPos, length);
+            ...
+          } else {
+            sN = peg$FAILED;
+            ...
+          }
+
+        compileInputChunkCondition will convert that to
+
+          sN = input.substr(peg$currPos, length);
+          if (<test>(sN)) {
+            ...
+          } else {
+            sN = peg$FAILED;
+            ...
+          }
+
+          and avoid extracting the sub string twice.
+      */
+      function compileInputChunkCondition(
+        cond, argCount, inputChunkLength
+      ) {
+        const baseLength = argCount + 3;
+        let inputChunk = inputChunkLength === 1
+          ? "input.charAt(peg$currPos)"
+          : "input.substr(peg$currPos, " + inputChunkLength + ")";
+        if (bc[ip + baseLength] === op.ACCEPT_N
+              && bc[ip + baseLength + 1] === inputChunkLength) {
+          skipAccept = true;
+          parts.push(stack.push(inputChunk));
+          inputChunk = stack.pop();
+        } else if (inputChunkLength === 1) {
+          cond = cond.replace("$$$.charCodeAt(0)", "input.charCodeAt(peg$currPos)");
+        }
+        compileCondition(
+          cond.replace("$$$", inputChunk),
+          argCount
+        );
```

Comment:
It would better just accept a transformation function instead of replacing special placeholder, something like `(inputChunk) => /*condition*/`